File: Lab1/cpp/searchMain.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<map>
#include<cmath>
#include<ctime>
// #include<string>
#include "search.cpp"

class LinearSearch{
public:
    std::map<int,std::clock_t> plottingDatas;
    int MAX,MIN,STEP;
    LinearSearch(){
        MAX = 200000;
        MIN = 10000;
        STEP = 5000;
        std::srand(time(NULL));
    }
    void clearData(){
        plottingDatas.clear();
    }
    void saveData(char* filename){
        FILE* f;
        f = fopen(filename,"w");
        fprintf(f,"# Number of Datas \t Time Interval (ms)\n");
        
        for(std::map<int,std::clock_t>::iterator it=plottingDatas.begin(); it!=plottingDatas.end();++it){
            fprintf(f,"%d\t%lf\n",it->first,(double)it->second);
            // std::cout<<"\t"<<it->first<<"\t"<<it->second<<std::endl;

        }
        std::cout<<"Data saved "<<filename<<std::endl;
        fclose(f);
    }
    void testBestCase(){
        std::vector<int> data;
        clearData();
        int index = 0;
        for(int i=MIN;i<MAX;i+=STEP){
            for(int j=0;j<i;j++){
                data.push_back(std::rand()%i);
            }
            std::clock_t start = std::clock();
            linearSearch(data,data[0]);
            std::clock_t end = std::clock();
            std::clock_t elapsed =(end-start);
            plottingDatas.insert({i,elapsed});
        }
        saveData("ls_best.dat");
    }
    void testAverageCase(){
        std::vector<int> data;
        clearData();

        int index = 0;
        for(int i=MIN;i<MAX;i+=STEP){
            for(int j=0;j<i;j++){
                data.push_back(std::rand()%i);
            }
            int r = data[rand()%i];
            std::clock_t start = std::clock();
            linearSearch(data,r);
            std::clock_t end = std::clock();
            std::clock_t elapsed = (end-start);
            // std::cout<<r<<" "<<elapsed<<std::endl;
            plottingDatas.insert({i,elapsed});
        }
        saveData("ls_avg.dat");
    }
    void testWorstCase(){
        std::vector<int> data;
        clearData();

        int index = 0;
        for(int i=MIN;i<MAX;i+=STEP){
            for(int j=0;j<i;j++){
                data.push_back(std::rand()%i);
            }
            int r = data[i-1];
            std::clock_t start = std::clock();
            linearSearch(data,r);
            std::clock_t end = std::clock();
            std::clock_t elapsed = (end-start);
            plottingDatas.insert({i,elapsed});
        }
        saveData("ls_worst.dat");
    }
    void plotData(){
        
        system("gnuplot ../plot.gp");
        std::cout<<"Data plotted, check your build file"<<std::endl;
    }
};


int main(){
    LinearSearch ls = LinearSearch();
    ls.testBestCase();
    ls.testAverageCase();
    ls.testWorstCase();
    ls.plotData();
    
    return 0;
}

File: Lab1/python/searchMain.py
from search import linear_search, binary_search
from time import time, time_ns
import matplotlib.pyplot as plt
import math

import random
import numpy as np


class LinearSearch:
    def __init__(self):
        self.plottingDatas = []
        self.MAX = 300000
        self.MIN = 10000
        self.STEP = 5000
        

    def clearData(self):
        self.plottingDatas = []

    def plotData(self, title,color):
        if(len(self.plottingDatas) <= 0):
            raise Exception("No data found")
        plt.plot(*zip(*self.plottingDatas),c=color)
        plt.title(title)
        plt.show()

    def testBestCase(self):
        print("Best Case")
        self.clearData()
        for length in range(self.MIN, self.MAX, self.STEP):
            data = random.sample(range(self.MAX), length)
            start = time()
            linear_search(data, data[0])
            end = time()
            elapsed = end-start
            print(f"Elapsed Time for {length} datas: {round(elapsed,6)}")
            self.plottingDatas.append((length, round(elapsed,6)))
        # print("Average Best Case: ",np.mean( list(zip(*self.plottingDatas))[1]))
        self.plotData("Linear Search- Best Case",'green')

    def testWorstCase(self):
        print("Worst Case")
        self.clearData()
        for length in range(self.MIN, self.MAX, self.STEP):
            data = random.sample(range(self.MAX), length)
            start = time()
            linear_search(data, data[-1])
            end = time()
            elapsed = end-start
            print(f"Elapsed Time for {length} datas: {round(elapsed,6)}")
            self.plottingDatas.append((length, round(elapsed,6)))
       
        # print("Average Worst Case: ",np.mean( list(zip(*self.plottingDatas))[1]))
      
        self.plotData("Linear Search- Worst Case",'blue')

    def testAverageCase(self):
        print("Average Case")
        self.clearData()
        for length in range(self.MIN, self.MAX, self.STEP):
            data = random.sample(range(self.MAX), length)
            r = random.randint(0, length)
            start = time()
            linear_search(data, data[r])
            end = time()
            elapsed = end-start
            print(f"Elapsed Time for {length} datas: {round(elapsed,6)}")
            self.plottingDatas.append((length, round(elapsed,6)))

        # print("Average Average Case: ",np.mean( list(zip(*self.plottingDatas))[1]))

        self.plotData("Linear Search- Average Case",'red')


class BinarySearch:
    def __init__(self):
        self.plottingDatas = []
        self.MAX = 10000000
        self.MIN = 10000
        self.STEP = 10000

    def clearData(self):
        self.plottingDatas = []

    def plotData(self, title):
        if(len(self.plottingDatas) <= 0):
            raise Exception("No data found")
        plt.plot(*zip(*self.plottingDatas))
        plt.title(title)
        plt.show()

    def testBestCase(self):
        print("Best Case")
        self.clearData()
        for length in range(self.MIN, self.MAX, self.STEP):
            data = range(length)
            start = time_ns()
            binary_search(data, data[length//2-1])
            end = time_ns()
            elapsed = end-start
            print(f"Elapsed Time for {length} datas: {round(elapsed,6)}")
            self.plottingDatas.append((length, round(elapsed,6)))

        # print("Average Best Case: ",np.mean( list(zip(*self.plottingDatas))[1]))

        self.plotData("Binary Search- Best Case")

    def testWorstCase(self):
        print("Worst Case")
        self.clearData()
        k = int(math.log2(self.MIN))+1
        n = 2 ** k - 1

        while(n <= self.MAX):
            data = range(n)
            start = time_ns()
            binary_search(data, data[-1])
            end = time_ns()
            elapsed = end-start
            print(f"Elapsed Time for {n} datas: {round(elapsed,6)}")
            self.plottingDatas.append((n, round(elapsed,6)))

            k += 1
            n = 2 ** k - 1

        # print("Average Worst Case: ",np.mean( list(zip(*self.plottingDatas))[1]))
        self.plotData("Binary Search- Worst Case")

    def testAverageCase(self):
        print("Average Case")
        self.clearData()
        for length in range(self.MIN, self.MAX, self.STEP):
            data = range(length)
            r = random.randint(0, length)
            start = time_ns()
            rand = data[r]
            # binary_search(data, data[r])
            binary_search(data, rand)
            end = time_ns()
            elapsed = end-start
            print(f"Elapsed Time for {length} datas: {round(elapsed,6)}")
            self.plottingDatas.append((length, round(elapsed,6)))

        # print("Average Average Case: ",np.mean( list(zip(*self.plottingDatas))[1]))
        self.plotData("Binary Search- aAverage Case")


if __name__ == "__main__":
    # LinearSearch().testAverageCase()
    # LinearSearch().testBestCase()
    LinearSearch().testWorstCase()
    # BinarySearch().testWorstCase()
    # BinarySearch().testBestCase()
    # BinarySearch().testAverageCase()
    # BinarySearch().testWorstCase()
